src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -67,19 +69,87 @@ async Task AuthenticateCleartext(string username, bool async)
 
         async Task AuthenticateSASL(List<string> mechanisms, string username, bool async)
         {
-            // At the time of writing PostgreSQL only supports SCRAM-SHA-256
-            if (!mechanisms.Contains("SCRAM-SHA-256"))
-                throw new NpgsqlException("No supported SASL mechanism found (only SCRAM-SHA-256 is supported for now). " +
+            // At the time of writing PostgreSQL only supports SCRAM-SHA-256 and SCRAM-SHA-256-PLUS
+            var supportsSha256 = mechanisms.Contains("SCRAM-SHA-256");
+            var supportsSha256Plus = mechanisms.Contains("SCRAM-SHA-256-PLUS");
+            if (!supportsSha256 && !supportsSha256Plus)
+                throw new NpgsqlException("No supported SASL mechanism found (only SCRAM-SHA-256 and SCRAM-SHA-256-PLUS are supported for now). " +
                                           "Mechanisms received from server: " + string.Join(", ", mechanisms));
-            var mechanism = "SCRAM-SHA-256";
+
+            var mechanism = string.Empty;
+            var cbindFlag = string.Empty;
+            var cbind = string.Empty;
+            var successfulBind = false;
+
+            if (supportsSha256Plus)
+            {
+                // RFC 5929
+                mechanism = "SCRAM-SHA-256-PLUS";
+                // PostgreSQL only supports tls-server-end-point binding
+                cbindFlag = "p=tls-server-end-point";
+                // SCRAM-SHA-256-PLUS depends on using ssl stream, so it's fine
+                var sslStream = (SslStream)_stream;
+                var cbindFlagBytes = Encoding.UTF8.GetBytes($"{cbindFlag},,");
+
+                using var remoteCertificate = new X509Certificate2(sslStream.RemoteCertificate);
+                // Checking for hashing algorithms
+                HashAlgorithm? hashAlgorithm = null;
+                var algorithmName = remoteCertificate.SignatureAlgorithm.FriendlyName;
+                if (algorithmName.IndexOf("sha1", StringComparison.OrdinalIgnoreCase) >= 0 ||
```

Comment:
I don't know anything about this, but is the algorithm name really not well-defined enough to make an actual comparison, rather than searching for strings inside it? Could the same algorithm name conceivably have more than one matching string?

Just asking questions, if this is appropriate for what you're seeing from PostgreSQL, I'm fine with it.

Reply:
It seems, PostgreSQL uses OpenSSL library for extracting hashing algorithm.
The main problem with all of this, is .net combines signing algorithm together with hashing algorithm. So, it looks like `sha1RSA`, `md5RSA` etc.
I think, we can change here to `StartsWith` instead of `IndexOf` and it would work just as well.

Reply:
You know better then me how this stuff works... I'm OK with whatever you think is right here.